src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs

```diff
@@ -71,109 +76,136 @@ private ParseUDFsResult ParseUDFs(string script)
                 {
-                    CreateError(_curs.TokCur, TexlStrings.ErrUDF_MissingEqual);
-                    break;
+                    TokEat(TokKind.Comma);
                 }
+            }
 
-                ParseTrivia();
+            ParseTrivia();
 
-                // Parse body
-                while (_curs.TokCur.Kind != TokKind.Semicolon)
-                {
-                    // Check if we're at EOF before a semicolon is found
-                    if (_curs.TidCur == TokKind.Eof)
-                    {
-                        CreateError(_curs.TokCur, TexlStrings.ErrUDF_MissingSemicolon);
-                        return new ParseUDFsResult(udfs, _errors);
-                    }
+            TokEat(TokKind.ParenClose);
+            return true;
+        }
+
+        private bool ParseUDF(List<UDF> udfs)
+        {
+            // <udf> ::= IDENT '(' <args> ')' ':' IDENT '=>' <function-body>
+            ParseTrivia();
+            var ident = TokEat(TokKind.Ident);
+            if (ident == null)
+            {
+                return false;
+            }
+
+            ParseTrivia();
+
+            if (!ParseUDFArgs(out HashSet<UDFArg> args))
+            {
+                return false;
+            }
+
+            ParseTrivia();
+
+            if (TokEat(TokKind.Colon) == null)
+            {
+                return false;
+            }
+
+            ParseTrivia();
+
+            var returnType = TokEat(TokKind.Ident);
+            if (returnType == null)
+            {
+                return false;
+            }
+
+            ParseTrivia();
 
-                    // Parse expression
-                    var result = ParseExpr(Precedence.None);
-                    udfs.Add(new UDF(ident.As<IdentToken>(), returnType.As<IdentToken>(), new HashSet<UDFArg>(args.Values), result));
+            if (TokEat(TokKind.DoubleBarrelArrow) == null)
+            {
+                return false;
+            }
+
+            // <function-body> ::= (EXP | <bracs-exp>)
+            // <bracs-exp> ::= '{' (((<EXP> ';')+ <EXP>) | <EXP>) (';')? '}'
+
+            ParseTrivia();
+
+            if (_curs.TidCur == TokKind.CurlyOpen)
+            {
+                _curs.TokMove();
+                _hasSemicolon = false;
+                ParseTrivia();
+                _flagsMode.Push(Flags.EnableExpressionChaining);
+                var exp_result = ParseExpr(Precedence.None);
+                ParseTrivia();
+                if (TokEat(TokKind.CurlyClose) == null)
+                {
+                    return false;
                 }
 
-                TokEat(TokKind.Semicolon);
+                udfs.Add(new UDF(ident.As<IdentToken>(), returnType.As<IdentToken>(), new HashSet<UDFArg>(args), exp_result, _hasSemicolon));
+                _flagsMode.Pop();
+                return true;
             }
 
-            return new ParseUDFsResult(udfs, _errors);
+            _flagsMode.Push(Flags.NamedFormulas);
```

Comment:
This is incorrect, you are pushing onto the stack 2 times

Reply:
It isn't, however the second push and pop group is unnecessary as it is simply is the same as the top level flags.